### PR TITLE
Properly handle :cq

### PR DESCRIPTION
**Pull request by bk2204**
When the user invokes `:cq`, this causes Neovim to exit unsuccessfully.
This is important if the user wishes to abort a Git commit message, or
in other situations where the operation is aborted when the editor exits
unsuccessfully.  However, neovim-gtk doesn't exit unsuccessfully in this case.

Let's make this work by listening to the `VimLeavePre` autocommand and
then reading the v:exiting code, which indicates the exit status with
which Neovim is exiting.  Save this into the state, and then when the
detach callback is called, exit with the appropriate status.

Note that we use a separate Arc/Mutex construction here because
otherwise we need to borrow the state mutably, and at that point it's
already borrowed and we panic.  This allows us to not need to do that,
and henceforth avoid the panic.

### CODE REVIEW

File: src/ui.rs [hunk 1]
```diff
@@ -259,14 +259,21 @@ impl Ui {
...
         window.connect_delete_event(clone!(comps_ref, shell_ref => move |_, _| {
             gtk_delete(&*comps_ref, &*shell_ref)
         }));
 
         shell.grab_focus();
 
-        shell.set_detach_cb(Some(clone!(comps_ref => move || {
+        shell.set_detach_cb(Some(clone!(comps_ref => move |status| {
             glib::idle_add_once(clone!(comps_ref => move || comps_ref.borrow().close_window()));
+            std::process::exit(status);
```
Comment (Lyude) [inline src/ui.rs:276]: Not sure this is the right way to go about this. Unlike the neovim-gtk of old, [we actually attempt to fully execute both `VimPreLeave` and `VimLeave` when exiting neovim-gtk](https://github.com/Lyude/neovim-gtk/blob/main/src/shell.rs#L1169) (which calls down to [`shutdown()`](https://github.com/Lyude/neovim-gtk/blob/main/src/nvim/mod.rs#L279) which I'm fairly sure an early exit here via `std::process::exit(…)` would skip. The main reason for doing this is that many plugins will have cleanup routines hooked up to these events, and exiting without executing them could result in things such as leaking plugin processes (YouCompleteMe was a good example of this).
Doing a bit of research, it seems the "proper" way of doing this would likely be to do it through the `GApplication` abstraction that GIO provides, since we use that in `src/main.rs`. [See here](https://gtk-rs.org/gtk-rs-core/stable/0.14/docs/gio/struct.ApplicationCommandLine.html), we would likely want to hook up to the command-line event, and use it to stash a copy of the `ApplicationCommandLine` into `Shell` that later from somewhere like `detach_ui()` we can pass the correct error code.

File: src/ui.rs [hunk 2]
```diff
@@ -558,7 +565,9 @@ fn on_help_about(window: &gtk::ApplicationWindow) {
     let about = AboutDialog::new();
     about.set_transient_for(Some(window));
     about.set_program_name("NeovimGtk");
-    about.set_version(Some(crate::GIT_BUILD_VERSION.unwrap_or(env!("CARGO_PKG_VERSION"))));
+    about.set_version(Some(
+        crate::GIT_BUILD_VERSION.unwrap_or(env!("CARGO_PKG_VERSION")),
+    ));
```
Comment (Lyude) [inline src/ui.rs:570]: Also, these changes don't seem related to the commit at hand

**Review by Lyude — CHANGES_REQUESTED** [review]
Thank you for the contribution! There's a couple of things that need to be fixed though

File: src/ui.rs (hunk 2 above)
Comment (bk2204) [reply src/ui.rs:570]: That was a `rustfmt`, I believe, but I can revert it.

Comment (Lyude) [reply src/ui.rs:570]: Ah that's fine then, I'd probably put it into a separate commit at least (can just be something short like "Rustfmt"

File: src/shell.rs [hunk 3]
```diff
@@ -435,6 +439,11 @@ impl State {
         }
     }
 
+    pub fn set_exit_status(&self, val: i32) {
+        let mut status = self.exit_status.lock().unwrap();
+        *status = Some(val);
```
[the remark this answers is not in the document]
Comment (bk2204) [reply src/shell.rs]: Fixed.

File: src/shell.rs [hunk 4]
```diff
@@ -1216,6 +1225,11 @@ impl Shell {
             .popup_menu
             .set_preview(options.contains("preview"));
     }
+
+    pub fn set_exit_status(&self, status: i32) {
+        let state = self.state.borrow();
+        state.set_exit_status(status);
```
[the remark this answers is not in the document]
Comment (bk2204) [reply src/shell.rs]: Done.

File: src/ui.rs (hunk 2 above)
Comment (bk2204) [reply src/ui.rs]: I've opted to just revert this because the code is not rustfmt clean at all and I've found that virtually every file will need to be touched.  We can do that in a separate PR.

File: src/ui.rs (hunk 1 above)
Comment (bk2204) [reply src/ui.rs]: I've gone ahead and used the `ApplicationCommandLine` class.  I noted that it seems to work differently than the documentation led me to believe, and perhaps it also works differently than you thought it worked, but I've documented that in the new commit message.

I've also gone ahead and made some relatively extensive changes to `main.rs` because those are required for us to use the new callback.  Again, most of this is documented in the commit message.

Our exit call is now at the end of the program, which I think is really the ideal place for it.

**Comment by Lyude** [comment]
Sorry for the slow reply! I've been off on vacation, will take a look at this sometime this week

**Comment by Lyude** [comment]
Changes look fine this time around, although it seems like there's still an issue with this: quitting nvim by closing the window and running with `--no-fork` causes it to spit this message out:

```
[2022-01-17T01:12:32Z ERROR nvim_gtk::nvim::handler] Error call function: Error reading args
```

Running in gdb, it seems like it's coming from:

```
#0  nvim_gtk::nvim::handler::safe_call::{closure#0}<nvim_gtk::nvim::handler::{impl#0}::nvim_cb::{generator#0}::{closure#
2}> () at src/nvim/handler.rs:235
#1  0x00005555558859fa in glib::source::fnmut_callback_wrapper::{closure#0}<nvim_gtk::nvim::handler::safe_call::{closure
#0}> () at /home/lyudess/.cargo/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/source.rs:159
#2  0x00005555558831bf in glib::source::trampoline<glib::source::fnmut_callback_wrapper::{closure#0}> (func=0x7fffd0000b
60) at /home/lyudess/.cargo/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/source.rs:90
#3  0x00007ffff719547b in g_idle_dispatch (source=0x7fffd0001060, callback=0x555555883180 <glib::source::trampoline<glib
::source::fnmut_callback_wrapper::{closure#0}>>, user_data=0x7fffd0000b60) at ../glib/gmain.c:5897
#4  0x00007ffff7199130 in g_main_dispatch (context=0x55555623d820) at ../glib/gmain.c:3381
#5  g_main_context_dispatch (context=0x55555623d820) at ../glib/gmain.c:4099
#6  0x00007ffff71ee208 in g_main_context_iterate.constprop.0 (context=context@entry=0x55555623d820, block=block@entry=0,
 dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/gmain.c:4175
#7  0x00007ffff7196933 in g_main_context_iteration (context=context@entry=0x55555623d820, may_block=may_block@entry=0) a
t ../glib/gmain.c:4240
#8  0x00007ffff73ba50c in g_application_run (application=0x55555629b210, argc=<optimized out>, argv=<optimized out>) at
../gio/gapplication.c:2600
#9  0x0000555555791ac7 in gio::application::{impl#0}::run_with_args<gtk::auto::application::Application, alloc::string::
String> (self=0x7fffffffd348, args=&[alloc::string::String](size=2) = {...}) at /home/lyudess/.cargo/registry/src/github
.com-1ecc6299db9ec823/gio-0.14.8/src/application.rs:30
#10 0x0000555555791bb6 in gio::application::{impl#0}::run<gtk::auto::application::Application> (self=0x7fffffffd348) at
/home/lyudess/.cargo/registry/src/github.com-1ecc6299db9ec823/gio-0.14.8/src/application.rs:23
#11 0x000055555574b78f in nvim_gtk::main () at src/main.rs:181
```

Which tells me it's probably a call to nvim failing in a `nvim::handler::safe_call` wrapper somewhere, likely from the connection for nvim being closed. I didn't dig deep enough to figure out why exactly we're managing to exit before redraw calls are finished. It might be that we need to make sure that we exit from `main.rs` while returning the status, without doing so before other threads are finished. Hopefully this doesn't end up having something to do with tokio :S.

**Comment by bk2204** [comment]
I've fixed that issue.  The only change is this:

```diff
         shell.state.borrow().subscribe(
             SubscriptionKey::from("VimLeave"),
-            &["v:exiting"],
+            &["v:exiting || 0"],
             clone!(shell_ref => move |args| set_exit_status(&*shell_ref, args)),
         );
```

When we close the window, we invoke the `VimLeavePre` and `VimLeave` autocommands.  In such a case, `v:exiting` is `v:null`, and without that change, we end up trying to stringify a Nil and failing.  Since we're closing the window, 0 is the right exit code to use in this case, and this should result in proper behavior without warnings.

**Comment by bk2204** [comment]
I've just fixed this from `"v:exiting || 0` to `v:exiting ? v:exiting : 0`, since I forgot that Vimscript has the C behavior for `||` (always 0 or 1) and not the Perl and Ruby behavior where one of the values is always selected.  That fixes `:cq 2` to exit 2.
